torch/fx/data_dependencies.py

```diff
@@ -0,0 +1,311 @@
+
+        # create new node
+        new_nodes[node_name] = graph.create_node(
+            op=op,
+            target=target,
+            args=new_args,
+            kwargs=new_kwargs,
+            name=orig_node.name,
+        )
+
+        if node_name in module_outputs:
+            graph_outputs.append(new_nodes[node_name])
+        if op == "placeholder":
+            graph_placeholders.append(new_nodes[node_name])
+
+    # set graph outputs 
+    if module_outputs:
+        graph.output(graph_outputs)
+    else:
+        graph.output(new_nodes[module_node_names[-1]])
+
+    # set up dict for copying over targets
+    subgraph_targets = {}
+    for node_name, node in new_nodes.items():
+        if node.op in ["get_attr", "call_module"]:
+            if node_name in submodule_hints:
+                # when node is a module created from this function, target is
+                # in orig_node's target directly
+                node_target = orig_nodes[node_name].target
+            else:
+                # the actual target still lives in the original torch.nn.module's hiearchy
+                node_target = orig_root_module
+                for atom in orig_nodes[node_name].target.split("."):
+                    node_target = getattr(node_target, atom)
+            subgraph_targets[str(node.target)] = node_target
+
+    module: torch.fx.GraphModule = torch.fx.GraphModule(subgraph_targets, graph)
+    module_as_node = torch.fx.Node(
+        graph=graph,
+        name=module_name,
+        op="call_module",
+        target=module,
+        args=tuple(graph_placeholders),
+        kwargs={},
+    )
+
+    return (module, module_as_node)
+
+
+class Partition:
+    def __init__(self, name: str):
+        self.name: str = name
+        self.nodes: List[str] = []
+        self.inputs: List[str] = []
+        self.outputs: List[str] = []
+        self.partitions_dependent_on: List[str] = []
+        self.partition_dependents: List[str] = []
+
+    def __repr__(self) -> str:
+        return f"name: {self.name},\n" \
+            f" nodes: {self.nodes},\n" \
+            f" inputs: {self.inputs},\n" \
+            f" outputs: {self.outputs},\n" \
+            f" partitions depenent on: {self.partitions_dependent_on},\n" \
+            f" parition dependents: {self.partition_dependents}"
+
+
+# Creates subgraphs out of main graph 
+def split_module(
+    m: torch.fx.GraphModule,
+    root_m: torch.nn.Module,
+    split_callback: Callable[[torch.fx.Node], int],
+):
+    partitions: Dict[str, Partition] = {}
+    orig_nodes: Dict[str, torch.fx.Node] = {}
+
+    # split nodes into parititons
+    for node in m.graph.nodes:
+        orig_nodes[node.name] = node
+
+        # TODO currently placeholders/parameters aren't put into random partitions, 
+        # rather they're added to the graphs where they are used down below
+        if node.op in ["placeholder", "get_attr"]:
+            continue
+        partition_name = split_callback(node)
```

Comment:
what do you think about having the split_callback() function return strings instead of ints? I was treating the return as a string in the rest of the code and it'll give people more freedom to name the submodules things like submod_gpu3 instead of submod_2

Reply:
Yeah makes sense to me